cmd/noms/noms_commit.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/attic-labs/noms/cmd/util"
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/dataset"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	flag "github.com/juju/gnuflag"
+)
+
+const (
+	dateFormat = "2006-01-02T15:04:05-0700"
+)
+
+var (
+	allowDupe bool
+	/*
+		meta      string
+		metaP     string
+	*/
+	message string
+	date    string
+)
+
+var nomsCommit = &util.Command{
+	Run:       runCommit,
+	UsageLine: "commit [options] <dataset> <path>",
+	Short:     "Commits to a dataset",
+	Long:      "If path is `--`, then the <path> is read from stdin. See Spelling Objects at https://github.com/attic-labs/noms/blob/master/doc/spelling.md for details on the dataset and path arguments.",
+	Flags:     setupCommitFlags,
+	Nargs:     2,
+}
+
+func setupCommitFlags() *flag.FlagSet {
+	commitFlagSet := flag.NewFlagSet("commit", flag.ExitOnError)
+	commitFlagSet.BoolVar(&allowDupe, "allow-dupe", false, "creates a new commit, even if it would be identical (modulo metadata and parents) to the existing HEAD.")
+	/*
+		commitFlagSet.StringVar(&meta, "meta", "", "'<key>:<value>' - creates a metadata field called 'key' set to 'value'. Value should be human-readable encoded.")
+		commitFlagSet.StringVar(&metaP, "meta-p", "", "'<key>:<path>' - creates a metadata field called 'key' set to the value at <path>")
+	*/
+	commitFlagSet.StringVar(&message, "message", "", "alias for -meta 'message:<message>'")
+	commitFlagSet.StringVar(&date, "date", "", "alias for -meta 'date:<date>'. '<date>' must be iso8601-formatted. If '<date>' is empty, it defaults to the current date.")
+	return commitFlagSet
+}
+
+func runCommit(args []string) int {
+	ds, err := spec.GetDataset(args[0])
+	d.CheckError(err)
+	defer ds.Database().Close()
+
+	pathDb, pathResolvedValue, err := spec.GetPath(args[1])
+	if err != nil {
+		d.CheckErrorNoUsage(err)
+	}
+	defer pathDb.Close()
+
+	if pathResolvedValue == nil {
+		d.CheckErrorNoUsage(fmt.Errorf("Object not found: %s", args[1]))
+	}
+
+	// Note that spec.GetPath() above gave us pathResolvedValue by calling
+	// ReadValue() already which loaded the cache for value into pathDb,
+	// but we are going to commit to ds.Database() so we need the cache for
+	// pathResolvedValue loaded into ds.Database() to perform the commit
+	// successfully
+	value := ds.Database().ReadValue(pathResolvedValue.Hash())
```

Comment:
Then in the code, you'd use `spec.GetDataset()`, then use `spec.NewAbsolutePath()` on the path, then call `Resolve()` on that against `ds.Database()`. Now everything is using the exact same database instance.